code/modules/music/tgui_panel/audio_extensions.dm

```diff
@@ -0,0 +1,135 @@
+		return FALSE
+	if(!track.audio._audio_asset && !findtext(track.audio._web_sound_url, GLOB.is_http_protocol))
+		return FALSE
+	if (track.audio._failed)
+		return FALSE
+	// Transport assets
+	if (track.audio._audio_asset)
+		track.audio._audio_asset.send(client)
+	var/list/payload = list()
+	var/list/extra_data = track.audio.get_additional_information()
+	if(length(extra_data) > 0)
+		for(var/key in extra_data)
+			payload[key] = extra_data[key]
+	if (payload["start"])
+		payload["start"] = payload["start"] + (world.time - track.started_at) * 0.1
+	else
+		payload["start"] = (world.time - track.started_at) * 0.1
+	if (istype(track.audio.license))
+		payload["license_title"] = track.audio.license.title
+		payload["license_url"] = track.audio.license.legal_text
+	payload["uuid"] = track.uuid
+	payload["url"] = track.audio._web_sound_url
+	payload["priority"] = track.priority
+	payload["flags"] = track.playing_flags
+	payload["volume"] = track.track_volume
+	window.send_message("audio/playMusic", payload)
+	return TRUE
+
+/**
+ * Play music with a given 3D position in the world
+ */
+/datum/tgui_panel/proc/play_spatial_music(datum/playing_track/spatial/track)
+	if(!is_ready())
+		return FALSE
+	if(!track.audio._audio_asset && !findtext(track.audio._web_sound_url, GLOB.is_http_protocol))
+		return FALSE
+	if (track.audio._failed)
+		return FALSE
+	// Transport assets
+	if (track.audio._audio_asset)
+		track.audio._audio_asset.send(client)
+	var/list/payload = list()
+	var/list/extra_data = track.audio.get_additional_information()
+	if(length(extra_data) > 0)
+		for(var/key in extra_data)
+			payload[key] = extra_data[key]
+	if (payload["start"])
+		payload["start"] = payload["start"] + (world.time - track.started_at) * 100
+	else
+		payload["start"] = (world.time - track.started_at) * 100
+	if (istype(track.audio.license))
+		payload["license_title"] = track.audio.license.title
+		payload["license_url"] = track.audio.license.legal_text
+	payload["uuid"] = track.uuid
+	payload["url"] = track.audio._web_sound_url
+	payload["priority"] = track.priority
+	var/turf/location = get_turf(track.source)
+	if (!location)
+		payload["x"] = 0
+		payload["y"] = 0
+		payload["z"] = 0
+	else
+		payload["x"] = location.x
+		payload["y"] = location.y
+		payload["z"] = location.z
+	payload["range"] = track.radius
+	payload["flags"] = track.playing_flags
+	payload["volume"] = track.track_volume
+	window.send_message("audio/playWorldMusic", payload)
+	needs_spatial_audio = TRUE
+	// Become a music listener
+	client.mob.AddComponent(/datum/component/music_listener, src)
+	return TRUE
+
+/datum/tgui_panel/proc/update_track_position(datum/playing_track/spatial/track)
+	if(!is_ready())
+		return FALSE
+	var/list/payload = list()
+	payload["uuid"] = track.uuid
+	var/turf/location = get_turf(track.source)
+	if (!location)
+		payload["x"] = 0
+		payload["y"] = 0
+		payload["z"] = 0
+	else
+		payload["x"] = location.x
+		payload["y"] = location.y
+		payload["z"] = location.z
```

Comment:
aren't we effectively giving clients a free GPS with this? should this not be a relative position that's sent as updates to the client rather than two absolute ones?

Reply:
There's not really an effective way to do this without absolute positions, as when someone moves it would require sending all the songs to them again relatively. Alternatively, we could only send the position of the active song but that would require logic on the server or communicating more between the clients. The cheapest and easiest way is to just send their position and accept that they might cheat to get GPS coordinates which isn't really too bad. I could add a random value between -1000 and 1000 to the positions on both ends to make it a little bit harder to use.

Reply:
I'd say you should just do the bare minimum computation on the server to turn the positions into relative ones and let the client do the rest with the volume. Doing basic subtraction is cheap.

Reply:
You already send every client their location, just send them their relative location to the nearest active jukebox

Reply:
This would be a significant amount more effort serverside. The risk is relatively low, but I'll take at least a low-security approach of generating random offsets for each round.